module/component/component.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// NewComponentManagerBuilder returns a new ComponentManagerBuilder
+func NewComponentManagerBuilder() ComponentManagerBuilder {
+	return &ComponentManagerBuilderImpl{
+		startup: func(ctx context.Context) error {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				return nil
+			}
+		},
+	}
+}
+
+func (c *ComponentManagerBuilderImpl) OnStart(startup ComponentStartup) ComponentManagerBuilder {
+	c.startup = startup
+	return c
+}
+
+func (c *ComponentManagerBuilderImpl) AddWorker(worker ComponentWorker) ComponentManagerBuilder {
+	c.workers = append(c.workers, worker)
+	return c
+}
+
+func (c *ComponentManagerBuilderImpl) AddComponent(component Component) ComponentManagerBuilder {
+	c.components = append(c.components, component)
+	return c
+}
+
+func (c *ComponentManagerBuilderImpl) Build() *ComponentManager {
+	return &ComponentManager{
+		started:     atomic.NewBool(false),
+		startupDone: make(chan struct{}),
+		ready:       make(chan struct{}),
+		startup:     c.startup,
+		workers:     c.workers,
+		components:  c.components,
+	}
+}
+
+var _ Component = (*ComponentManager)(nil)
+
+// ComponentManager is used to manage worker routines of a Component
+type ComponentManager struct {
+	started        *atomic.Bool
+	startupDone    chan struct{}
+	ready          chan struct{}
+	done           sync.WaitGroup
+	shutdownSignal <-chan struct{}
+
+	startup    func(context.Context) error
+	workers    []ComponentWorker
+	components []Component
+}
+
+// Start initiates the ComponentManager. It will first run the startup routine if one
+// was set, and then start all sub-components and launch all worker routines.
+func (c *ComponentManager) Start(parent irrecoverable.SignalerContext) (err error) {
+	if c.started.CAS(false, true) {
+		defer close(c.startupDone)
+
+		ctx, cancel := context.WithCancel(parent)
+		_ = cancel // pacify vet lostcancel check: startupCtx is always canceled through its parent
+		defer func() {
+			if err != nil {
+				cancel()
+			}
+		}()
+		c.shutdownSignal = ctx.Done()
+
+		if err = c.startup(ctx); err != nil {
+			return
+		}
+
+		signaler := irrecoverable.NewSignaler()
+		startupCtx := irrecoverable.WithSignaler(ctx, signaler)
+
+		go func() {
+			select {
+			case err := <-signaler.Error():
+				cancel()
+
+				// we propagate the error directly to the parent because a failure
+				// in a critical sub-component is considered irrecoverable
+				parent.Throw(err)
+			case <-ctx.Done():
+				// TODO: should we instead wait for parent.Done() or c.Done()?
+			}
+		}()
+
+		var componentGroup errgroup.Group
+		for _, component := range c.components {
+			component := component
+			componentGroup.Go(func() error {
+				if err := component.Start(startupCtx); err != nil {
+					defer cancel() // cancel startup for all other components
+					return err
+				}
+				return nil
+			})
+		}
+
+		if err = componentGroup.Wait(); err != nil {
+			// TODO: should we wait for c.Done() before returning?
```

Comment:
I think so. If `Start` returns an error, my expectation would be that I should throw it away and create a new one or panic()